Add initial Silabs SiM3U support

rettichschnidi opened the pull request:
This is the successor to #64209.

Noteworthy changes:
- moved to hwmv2, rebased on main
- using the official DK to make use of the SiM3U (adding the Gardena radio module will come later)
- put bindings in separate commits

Open questions:
- [x] <s>Would it be better to add the [picture provided by Silabs](https://silabs.scene7.com/is/image/siliconlabs/sim3u1xx-kit) instead of my (not too great) own shot?</s> Nobody complained, guess it is fine as is
- [x] Can/Should I add myself as maintainer? -> yes/done

TODO before merge:
 - [x] rebase once https://github.com/zephyrproject-rtos/hal_silabs/pull/66 is merged

Future:
 - Once this PR got merged, more drivers and support for the GARDENA smart garden radio module [will follow](https://github.com/husqvarnagroup/zephyr/commits/gardena/rs/sim3u/)

M1cha committed:
manifest: hal_silabs: Update HAL version

This bumps the hal_silabs in order to add initial support for SiM3U
SoCs.

Signed-off-by: Michael Zimmermann <michael.zimmermann@grandcentrix.net>

M1cha committed:
soc: Add initial SiM3U1xx support

This is the bare minimum and includes the SoC, pinctrl, flash and
devicetree.

I had to include the flash driver that early because I couldn't make
Zephyr compile without flash driver nodes in the device tree.

Signed-off-by: Michael Zimmermann <michael.zimmermann@grandcentrix.net>

M1cha committed:
drivers: clock_control: Add initial SiM3U1xx support

This serves two main purposes:
- change the CPU clock via devicetree nodes
- provide the APB frequency to device drivers via the clock driver
  interface

Theoretically this could also support choosing between the available
clock sources, but right now we only support LPOSC0 going into PLL0,
going into AHB.

Turning the PLL back off is also not supported since the only current
use case is to set the PLL frequency, turn it on, and switch the AHB
over to it.

Signed-off-by: Michael Zimmermann <michael.zimmermann@grandcentrix.net>

M1cha committed:
drivers: gpio: Add initial SiM3U1xx support

This is just the driver for banks 0 to 3. Bank 4 will come via a
separate commit since it needs a different driver.

Signed-off-by: Michael Zimmermann <michael.zimmermann@grandcentrix.net>

M1cha committed:
drivers: serial: Add initial SiM3U1xx support

This supports polling and interrupt APIs, but not the async API.

Signed-off-by: Michael Zimmermann <michael.zimmermann@grandcentrix.net>

rettichschnidi committed:
boards: silabs: Add SiM3U1xx-B-DK

SiM3U1xx-B-DK is the official development kit offered by Silicon Labs.

Signed-off-by: Reto Schneider <reto.schneider@husqvarnagroup.com>

rettichschnidi committed:
dts: bindings: Add Si32 flash controller

Initial version

Developed-by: Michael Zimmermann

Signed-off-by: Reto Schneider <reto.schneider@husqvarnagroup.com>

rettichschnidi committed:
dts: bindings: Add Si32 pinctrl

Initial version

Developed-by: Michael Zimmermann

Signed-off-by: Reto Schneider <reto.schneider@husqvarnagroup.com>

rettichschnidi committed:
drivers: flash: Add Si32 support

Can be used only once the SiM3U SoC support has been added.

Developed-by: Michael Zimmermann

Signed-off-by: Reto Schneider <reto.schneider@husqvarnagroup.com>

rettichschnidi committed:
dts: bindings: Add Si32 clocks

Initial version

Developed-by: Michael Zimmermann

Signed-off-by: Reto Schneider <reto.schneider@husqvarnagroup.com>

rettichschnidi committed:
dts: bindings: Add Si32 GPIO

Initial version

Developed-by: Michael Zimmermann

Signed-off-by: Reto Schneider <reto.schneider@husqvarnagroup.com>

rettichschnidi committed:
dts: bindings: Add Si32 UART

Initial version

Developed-by: Michael Zimmermann

Signed-off-by: Reto Schneider <reto.schneider@husqvarnagroup.com>

nordicjm commented on boards/silabs/dev_kits/sim3u1xx_dk/Kconfig.sim3u1xx_dk:
```diff
@@ -0,0 +1,7 @@
+# SPDX-License-Identifier: Apache-2.0
+
+config BOARD_SIM3U1XX_DK
+	bool "Silicon Labs SiM3U1xx 32-bit MCU USB Development Kit"
```
no selections

nordicjm commented on boards/silabs/dev_kits/sim3u1xx_dk/doc/index.rst:
```diff
@@ -0,0 +1,126 @@
++-------+--------------------------+----------------------------------------------------+
+|Pin    |Name                      | Note                                               |
++=======+==========================+====================================================+
+|PB1.12 | TX (O)                   | Serial connection to host via USB virtual COM port |
```
spaces around

nordicjm commented on boards/silabs/dev_kits/sim3u1xx_dk/doc/sim3u1xx_dk.jpg:
convert to webp and put through https://tinypng.com/

nordicjm commented on soc/silabs/Kconfig:
```diff
@@ -2,10 +2,10 @@
 
-if SOC_FAMILY_SILABS_S0 || SOC_FAMILY_SILABS_S1 || SOC_FAMILY_SILABS_S2
-
 rsource "*/Kconfig"
 
+if SOC_FAMILY_SILABS_S0 || SOC_FAMILY_SILABS_S1 || SOC_FAMILY_SILABS_S2
+
```
no

nordicjm replied:
why has this been broken again?

rettichschnidi replied:
Because of @jerome-pouiller's comment here: https://github.com/zephyrproject-rtos/zephyr/pull/76677#discussion_r1713586697

My understanding is that the way it is done now addresses his and your concern.

jerome-pouiller replied:
@nordicjm Silabs has 3 (now 4) SoC families and I would like to avoid complex conditions in `Kconfig`. I assume you would like to have the same pattern in every `soc/*/Kconfig` file, that's it?

Maybe the right pattern would be:
```
if SOC_FAMILY_SILABS_S0 || SOC_FAMILY_SILABS_S1 || SOC_FAMILY_SILABS_S2 || SOC_FAMILY_SILABS_SIM3

rsource "*/Kconfig"

if SOC_FAMILY_SILABS_S0 || SOC_FAMILY_SILABS_S1 || SOC_FAMILY_SILABS_S2
...
endif

endif
```

nordicjm replied:
No need to change

jerome-pouiller commented on soc/silabs/Kconfig:
```diff
@@ -318,4 +320,6 @@ config SOC_GECKO_CUSTOM_RADIO_PHY
-endif # SOC_FAMILY_SILABS_S0 || SOC_FAMILY_SILABS_S1 || SOC_FAMILY_SILABS_S2
+endif # !SOC_FAMILY_SILABS_SIM3
+
+endif # SOC_FAMILY_SILABS_S{0,1,2} || SOC_FAMILY_SILABS_SIM3
```
I am not super fond of all this boolean arithmetic. I believe it would be easier if we applied the condition in the inner Kconfig:
```
diff --git i/soc/silabs/Kconfig w/soc/silabs/Kconfig
index dd098de9559..c20fef18c9b 100644
--- i/soc/silabs/Kconfig
+++ w/soc/silabs/Kconfig
@@ -2,10 +2,10 @@
 # Copyright (c) 2018 Gil Benkoe
 # SPDX-License-Identifier: Apache-2.0

-if SOC_FAMILY_SILABS_S0 || SOC_FAMILY_SILABS_S1 || SOC_FAMILY_SILABS_S2
-
 rsource "*/Kconfig"

+if SOC_FAMILY_SILABS_S0 || SOC_FAMILY_SILABS_S1 || SOC_FAMILY_SILABS_S2
+
 config SOC_GECKO_SDID
        int
        help
diff --git i/soc/silabs/silabs_s0/Kconfig w/soc/silabs/silabs_s0/Kconfig
index dfc253eb424..6baa33bc1ac 100644
--- i/soc/silabs/silabs_s0/Kconfig
+++ w/soc/silabs/silabs_s0/Kconfig
@@ -2,12 +2,12 @@
 #
 # SPDX-License-Identifier: Apache-2.0

+if SOC_FAMILY_SILABS_S0
+
 config SOC_FAMILY_SILABS_S0
        select HAS_SEGGER_RTT if ZEPHYR_SEGGER_MODULE
        select BUILD_OUTPUT_HEX

-if SOC_FAMILY_SILABS_S0
-
 rsource "*/Kconfig"

 endif # SOC_FAMILY_SILABS_S0
...
```

rettichschnidi replied:
Adapted accordingly (I think).

jerome-pouiller reviewed (commented):
Beside my small comment, the code looks good to me.

I think you should add your name as maintainer of this code.

rettichschnidi committed:
MAINTAINERS: Add SiM3U maintainer and collaborators

Adding myself as maintainer because my employer Husqvarna wants to
upstream SiM3U support to Zephyr, and keep it well-supported.

Adding as collaborator M1cha, because he did the initial port and
Asmellby, jerome-pouiller, and jhedberg because they take care of
all Silabs products.

Signed-off-by: Reto Schneider <reto.schneider@husqvarnagroup.com>

rettichschnidi committed:
soc: silabs: Simplify logic

This makes it easier to add non-gecko SoCs.

Signed-off-by: Reto Schneider <reto.schneider@husqvarnagroup.com>

jhedberg commented:
@rettichschnidi please update the `west.yml` reference now that the HAL PR has been merged. Thanks.

rettichschnidi commented:
> @rettichschnidi please update the `west.yml` reference now that the HAL PR has been merged. Thanks.

Done.

Also done:
 - Rebased on most recent main
 - Fixed two (small) GPIO and flash issues, so that `west twister --platform sim3u1xx_dk --device-testing --hardware-map my-hardware-map.yml` passes all tests
 - Added @M1cha (he did the initial port)

jhedberg commented on MAINTAINERS.yml:
```diff
@@ -3490,6 +3490,24 @@ Silabs Platforms:
+  status: maintained
+  maintainers:
+    - rettichschnidi
+  collaborators:
+    - M1cha
```
Could you also add the maintainers & collaborators of Silabs platforms in general (@jhedberg @asmellby and @jerome-pouiller) to collaborators list here? That'll make sure that we get added as reviewers to relevant issues and pull requests.

rettichschnidi replied:
done

jhedberg commented on MAINTAINERS.yml:
```diff
@@ -3490,6 +3490,24 @@ Silabs Platforms:
+    - dts/arm/silabs/sim3u*
+    - dts/bindings/*/*silabs,si32*
+    - soc/silabs/silabs_sim3/
+  labels:
+    - "platform: Silabs SiM3U"
```
This label will need to be separately created (doesn't seem to exist right now). I can do that.
